npm/webpack-dev-server/src/plugin.ts

```diff
@@ -52,85 +53,81 @@ export default class CypressCTOptionsPlugin {
   }
 
-  private pluginFunc = (context: CypressCTWebpackContext) => {
-    context._cypress = {
+  private addLoaderContext = (loaderContext: object, module: any) => {
+    (loaderContext as CypressCTWebpackContext)._cypress = {
       files: this.files,
       projectRoot: this.projectRoot,
       supportFile: this.supportFile,
     }
   };
 
-  private setupCustomHMR = (compiler: webpack.Compiler) => {
-    compiler.hooks.afterCompile.tap(
-      'CypressCTOptionsPlugin',
-      (compilation) => {
-        const stats = compilation.getStats()
-
-        if (stats.hasErrors()) {
-          this.errorEmitted = true
-
-          // webpack 4: string[]
-          // webpack 5: Error[]
-          const errors = stats.toJson().errors as Array<Error | string> | undefined
-
-          if (!errors || !errors.length) {
-            return
-          }
-
-          this.devServerEvents.emit('dev-server:compile:error', normalizeError(errors[0]))
-        } else if (this.errorEmitted) {
-          // compilation succeed but assets haven't emitted to the output yet
-          this.devServerEvents.emit('dev-server:compile:error', null)
-        }
-      },
-    )
-
-    compiler.hooks.afterEmit.tap(
-      'CypressCTOptionsPlugin',
-      (compilation) => {
-        if (!compilation.getStats().hasErrors()) {
-          this.devServerEvents.emit('dev-server:compile:success')
-        }
-      },
-    )
+  private afterCompile = () => {
+    if (!this.compilation) {
+      return
+    }
+
+    const stats = this.compilation.getStats()
+
+    if (stats.hasErrors()) {
+      this.errorEmitted = true
+
+      // webpack 4: string[]
+      // webpack 5: Error[]
+      const errors = stats.toJson().errors as Array<Error | string> | undefined
+
+      if (!errors || !errors.length) {
+        return
+      }
+
+      this.devServerEvents.emit('dev-server:compile:error', normalizeError(errors[0]))
+    } else if (this.errorEmitted) {
+      // compilation succeed but assets haven't emitted to the output yet
+      this.devServerEvents.emit('dev-server:compile:error', null)
+    }
+  }
+
+  private afterEmit = () => {
+    if (!this.compilation?.getStats().hasErrors()) {
+      this.devServerEvents.emit('dev-server:compile:success')
+    }
+  }
+
+  private onSpecsChange = (specs: Cypress.Cypress['spec'][]) => {
+    if (!this.compilation || _.isEqual(specs, this.files)) {
+      return
+    }
+
+    this.files = specs
+    const inputFileSystem = this.compilation.inputFileSystem
+    const utimesSync: UtimesSync = semver.gt('4.0.0', webpack.version) ? inputFileSystem.fileSystem.utimesSync : fs.utimesSync
+
+    utimesSync(path.resolve(__dirname, 'browser.js'), new Date(), new Date())
   }
 
   /**
    *
    * @param compilation webpack 4 `compilation.Compilation`, webpack 5
    *   `Compilation`
    */
-  private plugin = (compilation: Webpack45Compilation) => {
-    this.devServerEvents.on('dev-server:specs:changed', (specs) => {
```

Comment:
This listener was getting added every compilation, and never removed. Since it holds onto a reference to `compilation`, it was also a memory leak (though not a severe one) in addition to a listener lead.